File: microservices/categories-service/src/main/java/com/categoriesservice/categories/controller/CategoryController.java
package com.categoriesservice.categories.controller;

import com.categoriesservice.categories.dto.CategoryDTO;
import com.categoriesservice.categories.service.CategoryService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * Category Controller - REST API para gestión de categorías
 */
@RestController
@RequestMapping("/api/v1/categories")
public class CategoryController {
    
    @Autowired
    private CategoryService categoryService;
    
    @GetMapping
    public ResponseEntity<List<CategoryDTO>> listarTodas() {
        return ResponseEntity.ok(categoryService.listarTodas());
    }
    
    @PostMapping
    public ResponseEntity<CategoryDTO> crear(@RequestBody CategoryDTO categoryDTO) {
        CategoryDTO nuevaCategory = categoryService.guardar(categoryDTO);
        return ResponseEntity.ok(nuevaCategory);
    }
    
    @GetMapping("/{id}")
    public ResponseEntity<CategoryDTO> obtenerPorId(@PathVariable Long id) {
        return categoryService.obtenerPorId(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }
    
    @GetMapping("/user/{userId}")
    public ResponseEntity<List<CategoryDTO>> obtenerPorUsuario(@PathVariable Long userId) {
        return ResponseEntity.ok(categoryService.obtenerPorUsuario(userId));
    }
    
    @GetMapping("/search")
    public ResponseEntity<List<CategoryDTO>> buscarPorNombre(@RequestParam String name) {
        return ResponseEntity.ok(categoryService.buscarPorNombre(name));
    }
    
    @PutMapping("/{id}")
    public ResponseEntity<CategoryDTO> actualizar(@PathVariable Long id, @RequestBody CategoryDTO categoryDTO) {
        return categoryService.actualizar(id, categoryDTO)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }
    
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> eliminar(@PathVariable Long id) {
        if (categoryService.eliminar(id)) {
            return ResponseEntity.ok().build();
        }
        return ResponseEntity.notFound().build();
    }
}

File: microservices/categories-service/src/main/kotlin/com/categoriesservice/categories/controller/CategoryController.kt
package com.categoriesservice.categories.controller

import com.categoriesservice.categories.dto.CategoryDTO
import com.categoriesservice.categories.service.CategoryService
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/categories")
class CategoryController(
    private val service: CategoryService
) {
    
    @GetMapping
    fun getAllCategories(): ResponseEntity<List<CategoryDTO>> =
        ResponseEntity.ok(service.findAll())
    
    @GetMapping("/active")
    fun getActiveCategories(): ResponseEntity<List<CategoryDTO>> =
        ResponseEntity.ok(service.findAllActive())
    
    @GetMapping("/{id}")
    fun getCategoryById(@PathVariable id: Long): ResponseEntity<CategoryDTO> =
        ResponseEntity.ok(service.findById(id))
    
    @PostMapping
    fun createCategory(@Valid @RequestBody dto: CategoryDTO): ResponseEntity<CategoryDTO> {
        val created = service.create(dto)
        return ResponseEntity.status(HttpStatus.CREATED).body(created)
    }
    
    @PutMapping("/{id}")
    fun updateCategory(
        @PathVariable id: Long,
        @Valid @RequestBody dto: CategoryDTO
    ): ResponseEntity<CategoryDTO> {
        val updated = service.update(id, dto)
        return ResponseEntity.ok(updated)
    }
    
    @DeleteMapping("/{id}")
    fun deleteCategory(@PathVariable id: Long): ResponseEntity<Void> {
        service.delete(id)
        return ResponseEntity.noContent().build()
    }
    
    @PatchMapping("/{id}/toggle-active")
    fun toggleActive(@PathVariable id: Long): ResponseEntity<CategoryDTO> {
        val updated = service.toggleActive(id)
        return ResponseEntity.ok(updated)
    }
}
